### PR TITLE
FOUNDATIONS: Add Team

### LINKED ISSUE #61: FOUNDATIONS: Insert `Team`

### PR DESCRIPTION
**Pull request by DadabaevAlisher**
Closes #61

### CODE REVIEW

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                return await returningTeamFunction();
+            }
+            catch (NullTeamException nullTeamException)
+            {
+                throw CreateAndLogValidationException(nullTeamException);
+            }
+            catch(InvalidTeamException invalidTeamException)
+            {
+                throw CreateAndLogValidationException(invalidTeamException);
+            }catch(SqlException sqlException)
```
[the remark this answers is not in the document]
Comment (DadabaevAlisher) [reply Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs]: Done

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                var failedTeamStorageException = new FailedTeamStorageException(sqlException);
+
+                throw CreateAndLogValidationException(sqlException);
+            }
+        }
+
+        private Exception CreateAndLogValidationException(SqlException sqlException)
+        {
+            throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs]: Done

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        public TeamService(
+            IStorageBroker storageBroker,
+            ILoggingBroker loggingBroker)
+        {
+            this.storageBroker = storageBroker;
+            this.loggingBroker = loggingBroker;
+        }
+
+        public ValueTask<Team> AddTeamAsync(Team team) =>
+            TryCatch(async () =>
```
[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api/Services/Foundations/Teams/TeamService.cs]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            var lockedTeamException = new LockedTeamException(dbUpdateConcurrencyException);
+
+            var expectedTeamDependencyValidationException = 
+                new TeamDependencyValidationException(lockedTeamException);
+
+            this.storageBrokerMock.Setup(broker=>broker.InsertTeamAsync(It.IsAny<Team>()))
+                .ThrowsAsync(dbUpdateConcurrencyException);
+            
+            //when
+            ValueTask<Team> addTeamTask = this.teamService.AddTeamAsync(someTeam);
```
Comment (Javlonbek-dev) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:109]: Add Line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Add.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                broker.GetCurrentDateTime()).Returns(randomDateTime);
+
+            //when
+            ValueTask<Team> addTeamTask = this.teamService.AddTeamAsync(invalidTeam);
+
+            TeamValidationException actualTeamValidationException =
+                await Assert.ThrowsAsync<TeamValidationException>(addTeamTask.AsTask);
+
+            // then
+
```
Comment (Javlonbek-dev) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Add.cs:170]: Remove Line

File: Tarteeb.Api/Models/Teams/Team.cs [hunk 6]
```diff
@@ -13,5 +13,6 @@ public class Team
         public string TeamName { get; set; }
         public DateTimeOffset CreatedDate { get; set; }
         public DateTimeOffset UpdatedDate { get; set; }
+        public Guid UpdatedUserId { get; set; }
```
Comment (Javlonbek-dev) [inline Tarteeb.Api/Models/Teams/Team.cs:16]: we don't need it

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                .BeEquivalentTo(expectedTeamDependencyValidationException);
+
+            this.storageBrokerMock.Verify(broker => broker.InsertTeamAsync(It.IsAny<Team>()),Times.Once);
+
+            this.loggingBrokerMock.Verify(broker => broker.LogError(It.Is(
+                SameExceptionAs(expectedTeamDependencyValidationException))), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+
```
Comment (Javlonbek-dev) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:124]: Remove Line

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            ValidateTeamNotNull(team);
+
+            Validate(
+                (Rule: IsInvalid(team.Id), Parameter: nameof(Team.Id)),
+                (Rule: IsInvalid(team.TeamName), Parameter: nameof(Team.TeamName)),
+                (Rule: IsInvalid(team.CreatedDate), Parameter: nameof(Team.CreatedDate)),
+                (Rule: IsInvalid(team.UpdatedDate), Parameter: nameof(Team.UpdatedDate)),
+                (Rule: IsInvalid(team.UpdatedUserId), Parameter: nameof(Team.UpdatedUserId)),
+                (Rule: IsNotSame(team.CreatedDate, team.UpdatedDate, nameof(Team.UpdatedDate)),
```
Comment (Javlonbek-dev) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:25]: firstDate:team. CreateDate, secondDate:team. UpdatedDate

Comment (Javlonbek-dev) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:25]: firstDate:team. CreateDate, secondDate:team. UpdatedDate

Comment (DadabaevAlisher) [reply Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:24]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs (hunk 7 above)
Comment (DadabaevAlisher) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Add.cs (hunk 5 above)
Comment (DadabaevAlisher) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Add.cs]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs (hunk 4 above)
Comment (DadabaevAlisher) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:109]: Done

File: Tarteeb.Api/Models/Teams/Team.cs (hunk 6 above)
Comment (StudentofTATU) [reply Tarteeb.Api/Models/Teams/Team.cs:16]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Add.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                .AddTeamAsync(inputTeam);
+
+            //then
+            actuaTeam.Should().BeEquivalentTo(expectedTeam);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertTeamAsync(inputTeam), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Add.cs:50]: Done

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        private readonly Mock<IStorageBroker> storageBrokerMock;
+        private readonly Mock<ILoggingBroker> loggingBrokerMock;
+        private readonly ITeamService teamService;
+
+        public TeamServiceTests()
+        {
+            this.storageBrokerMock = new Mock<IStorageBroker>();
+            this.loggingBrokerMock = new Mock<ILoggingBroker>();
+
```
[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs:34]: Done

File: Tarteeb.Api/Models/Teams/Exceptions/FailedTeamServiceException.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+using System;
+using Xeptions;
+
+namespace Tarteeb.Api.Models.Teams.Exceptions
+{
+    public class FailedTeamServiceException : Xeption
+    {
+        public FailedTeamServiceException(Exception innerException)
+            : base(message: "Failed profile service occurred, please contact support.",
+                  innerException)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Models/Teams/Exceptions/FailedTeamServiceException.cs:15]: tab

File: Tarteeb.Api/Models/Teams/Exceptions/TeamServiceException.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+using System;
+using Xeptions;
+
+namespace Tarteeb.Api.Models.Teams.Exceptions
+{
+    public class TeamServiceException : Xeption
+    {
+        public TeamServiceException(Exception innerException)
+            : base(message: "Profile service error occurred, contact support.", innerException)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Models/Teams/Exceptions/TeamServiceException.cs:14]: `Profile` ?

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            return teamValidationExpcetion;
+        }
+
+        private TeamDependencyException CreateAndLogCriticalDependencyException(Xeption exception)
+        {
+            var teamDependencyException = new TeamDependencyException(exception);
+            this.loggingBroker.LogCritical(teamDependencyException);
+
+            return teamDependencyException;
+        }
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:75]: Bo'sh katak tashlang

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+            var expectedTeamDependencyValidationException =
+                new TeamDependencyValidationException(lockedTeamException);
+
+            this.dateTimeBrokerMock.Setup(broker => broker.GetCurrentDateTime())
+                .Throws(dbUpdateConcurrencyException);
+
+            //when
+            ValueTask<Team> addTeamTask = this.teamService.AddTeamAsync(someTeam);
+
+            var actualTeamDependencyValidationException =
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:117]: `var` o'rniga aniq tiping yozing

File: Tarteeb.Api/Models/Teams/Exceptions/FailedTeamServiceException.cs (hunk 11 above)
Comment (StudentofTATU) [reply Tarteeb.Api/Models/Teams/Exceptions/FailedTeamServiceException.cs:15]: Done

**Review by jamshidbek001 — APPROVED** [review]

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+        [Fact]
+        public async Task ShouldTrowDependencyValidationExceptionOnAddIfDuplicateErrorOccursAndLogItAsync()
+        {
+            // given
+            Team someTeam = CreateRandomTeam();
+            string someMessage = GetRandomString();
+            var duplicateKeyException = new DuplicateKeyException(someMessage);
+
+            var alreadyExistsTicketException =
```
[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:67]: I think  'var' is right

[the remark this answers is not in the document]
Comment (Javlonbek-dev) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs:67]: Ha toʻgʻri lekin Ticketni Teamga oʻzgartiring

[the remark this answers is not in the document]
Comment (StudentofTATU) [reply Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Add.cs]: Done

**Review by jamshidbek001 — APPROVED** [review]

**Pull request merged**